Feat: run vcluster as non root user

matskiv opened the pull request:
**Major changes:**
Instead of letting k3s configure CoreDNS, we install it into vcluster from manifests/coredns/coredns.yaml,
which is built into a container image and is applied on each syncer start.
This approach also allows a user to mount an alternative coredns manifest from a Configmap.

The contents of coredns.yaml are taken from k3s with only minor changes (port 1053 instead of 53, different templating syntax, templating securityContext.
A new controller keeps NodeHosts values of the CoreDNS config map up to date.

~~Vcluster kube config is written into a new location, we can't use /root/ now, but the old location is symlinked to support the old cli versions.~~
Edit: Vcluster kube config is now always written into the default secret, and the related flag creates an additional kube config secret. For backward compatibility with vcluster CLI vcluster attempts to write kube config into a file as it did before, but ignores any errors.

Additional optional Role(Binding) has been added for installation on OpenShift. It's creation is enabled with helm value `openshift.enable`.

For additional security, the data volume is mounted as read-only.
Certificate syncer has been refactored to not write into a file, this allows users to set `securityContext.readOnlyRootFilesystem=true`.

**Other:**
In the builder image, which is also used for development, the /manifests location is symlinked to synced folder.
Dockerfile contains small improvements to dev build speed (GOCACHE).
Tooling for local development has been slightly improved.
Devspace profiles now include dev-crc and deploy-crc for usage with local OpenShift 4 instance (requires CRC and Podman).

**Warning:**
A Vcluster upgrade of an instance that uses root user into one that is not using it has not been tested, and is considered unsupported for now.

Fixes #67 
Fixes #171 
Fixes #189

matskiv commented on chart/values.yaml:
```diff
@@ -119,3 +122,16 @@ ingress:
+# Configure SecurityContext of the VCluster pod
+securityContext:
+  allowPrivilegeEscalation: false
+  # readOnlyRootFilesystem: true
```
I meant to add a comment here^ that would link to #189,
readOnlyRootFilesystem: true doesn't work right now.

I'll add that comment if I'll be doing any follow-up fixes or changes.

matskiv commented:
After a discussion with @FabianKramm I did some changes around kube config handling, etc.
The updated code is now pushed and is ready for review and testing.
The only remaining bit is a new e2e test for the DNS setup.

FabianKramm commented on pkg/server/cert/cert.go:
```diff
@@ -3,14 +3,15 @@ package cert
 )
 
-func GenServingCerts(caCertFile, caKeyFile, certFile, keyFile, clusterDomain string, SANs []string) (bool, error) {
+func GenServingCerts(caCertFile, caKeyFile string, currentCert, currentKey *[]byte, clusterDomain string, SANs []string) (bool, error) {
```
please return new key and certificate instead of replacing them in-place in the parameters

FabianKramm commented on pkg/server/cert/syncer.go:
```diff
@@ -3,27 +3,25 @@ package cert
 
 var (
-	certPath = "/var/lib/vcluster/tls"
+	certPath = "/data/var/lib/vcluster/tls"
```
is this still needed?

matskiv replied:
Nope, this is unused. I'll delete it.

FabianKramm reviewed (changes requested):
@matskiv thanks for the PR! Please see my comments.

matskiv commented on chart/values.yaml:
```diff
@@ -119,3 +120,23 @@ ingress:
+  # To run vcluster pod as non-root uncomment runAsUser and runAsNonRoot values.
+  # Update the runAsUser value if your cluster has limitations on user UIDs.
+  # For installation on OpenShift leave the runAsUser undefined (commented out).
+  runAsUser: 12345
```
Oh yeah, this was changed for testing, I forgot to comment it out again. Thanks for noticing :+1:

matskiv commented on pkg/coredns/coredns.go:
```diff
@@ -0,0 +1,118 @@
+)
+
+var CoreDNSVersionMap = map[string]string{
+	"1.22": "rancher/mirrored-coredns-coredns:1.8.4",
```
Yes, I forgot to update this map.
I'll test it on a few previous versions and commit the update.

matskiv commented on pkg/util/certhelper/cert.go:
```diff
@@ -110,7 +75,7 @@ func NewSignedCert(cfg Config, key crypto.Signer, caCert *x509.Certificate, caKe
 		ExtKeyUsage:  cfg.Usages,
 	}
-	certDERBytes, err := x509.CreateCertificate(cryptorand.Reader, &certTmpl, caCert, key.Public(), caKey)
+	certDERBytes, err := x509.CreateCertificate(rand.Reader, &certTmpl, caCert, key.Public(), caKey)
```
It's the same package, just imported second time under "cryptorand" name:
https://github.com/loft-sh/vcluster/blob/c09ff5de283fd313b47606f7d6b38d023839453e/pkg/util/certhelper/cert.go#L21-L22

matskiv commented:
I addressed the comments + added two e2e tests and made them run testing pods as non-root.

I also added a commit with a fallback code for the connect command. Please check the last commit in this PR and let me know if we should keep this, or whether it's better to keep the code clean despite breaking backward compatibility (if vcluster is using a different value of the `--out-kube-config-secret` flag).

matskiv commented on CONTRIBUTING.md:
```diff
@@ -110,9 +110,8 @@ If you are not using VSCode, configure your IDE to connect to `localhost:2346` f
 
-The e2e test suite can be run from the e2e folder(`cd e2e`) with this command - `VCLUSTER_SUFFIX=vcluster go test -v -ginkgo.v`.  
-Alternatively, if you [install ginkgo binary](https://github.com/onsi/ginkgo#global-installation), you can run it with `VCLUSTER_SUFFIX=vcluster ginkgo -v`.  
-If you are running vcluster with `devspace run dev` then the `VCLUSTER_SUFFIX` environment variable shouldn't be set, and you should run the tests with `go test -v -ginkgo.v` or `ginkgo -v`.  
```
FYI: I remove these mentions of the `VCLUSTER_SUFFIX` env var because the syncer and e2e test suite will now default to `vcluster` value for the suffix, which is the same as we've been using in the helm chart, so setting the env var is no longer necessary for running e2e tests.

FabianKramm reviewed (approved):
@matskiv thanks a lot! LGTM!
